### PR TITLE
fix deparse of state sub with empty prototype

### LINKED ISSUE #21315: Can't deparse state constant function definition
Module: B::Deparse

**Description**

```
$ perl -MO=Deparse -Mfeature=state -e 'state sub FOO :prototype() { 42 } print(FOO, "\n")'
While deparsing -e near line 1,
Can't locate object method "object_2svref" via package "B::SPECIAL" at /usr/share/perl/core/B/Deparse.pm line 466.
CHECK failed--call queue aborted.
```

**Steps to Reproduce**

`perl -MO=Deparse -Mfeature=state -e 'state sub FOO :prototype() { 42 } print(FOO, "\n")'`

**Expected behavior**

Something akin to `my` constant function definition:
```
perl -MO=Deparse -e 'my sub FOO :prototype() { 42 } print(FOO, "\n")'
my sub FOO () { 42 }

print 42, "\n";
-e syntax OK
```

**Perl configuration**

```
Summary of my perl5 (revision 5 version 38 subversion 0) configuration:

  Platform:
    osname=msys
    osvers=3.4.7.x86_64
    archname=x86_64-msys
    uname='msys_nt-10.0-19045 desktop-97pareg 3.4.7.x86_64 2023-07-14 16:57 utc x86_64 msys '
    config_args='-des -Uusethreads -Doptimize=-march=nocona -msahf -mtune=generic -O2 -pipe -Dprefix=/usr -Dprivlib=/usr/share/perl/core -Darchlib=/usr/lib/perl/core -Dsitelib=/usr/share/perl/site -Dsitearch=/usr/lib/perl/site -Dinc_version_list=none -Dman1dir=none -Dman3dir=none -Darchname=-msys -Dmyarchname=-msys -Dlibperl=msys-perl5_38.dll -Dcc=gcc -Dld=g++ -Accflags=-march=nocona -msahf -mtune=generic -O2 -pipe -DNO_TAINT_SUPPORT -fwrapv'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=undef
    usemultiplicity=undef
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='gcc'
    ccflags ='-DPERL_USE_SAFE_PUTENV -U__STRICT_ANSI__ -D_GNU_SOURCE -march=nocona -msahf -mtune=generic -O2 -pipe -DNO_TAINT_SUPPORT -fwrapv -fno-strict-aliasing -fstack-protector-strong -D_FORTIFY_SOURCE=2'
    optimize='-march=nocona -msahf -mtune=generic -O2 -pipe'
    cppflags='-DPERL_USE_SAFE_PUTENV -U__STRICT_ANSI__ -D_GNU_SOURCE -march=nocona -msahf -mtune=generic -O2 -pipe -DNO_TAINT_SUPPORT -fwrapv -fno-strict-aliasing -fstack-protector-strong'
    ccversion=''
    gccversion='11.3.0'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='g++'
    ldflags =' -Wl,--enable-auto-import -Wl,--export-all-symbols -Wl,--enable-auto-image-base -fstack-protector-strong'
    libpth=/usr/lib
    libs=-lpthread -lgdbm -ldb -ldl -lcrypt -lgdbm_compat
    perllibs=-lpthread -ldl -lcrypt
    libc=/usr/lib/libmsys-2.0.a
    so=dll
    useshrplib=true
    libperl=msys-perl5_38.dll
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=dll
    d_dlsymun=undef
    ccdlflags=' '
    cccdlflags=' '
    lddlflags=' --shared  -Wl,--enable-auto-import -Wl,--export-all-symbols -Wl,--enable-auto-image-base -fstack-protector-strong'


Characteristics of this binary (from libperl):
  Compile-time options:
    HAS_LONG_DOUBLE
    HAS_STRTOLD
    HAS_TIMES
    NO_TAINT_SUPPORT
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_HASH_FUNC_SIPHASH13
    PERL_HASH_USE_SBOX32
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    PERL_USE_SAFE_PUTENV
    USE_64_BIT_ALL
    USE_64_BIT_INT
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_PERLIO
    USE_PERL_ATOF
  Built under msys
  Compiled at Jul 28 2023 13:55:54
  %ENV:
    PERLLIB="/home/iv/.my/perl"
    MSYS="winsymlinks:sys"
  @INC:
    /home/iv/.my/perl
    /usr/lib/perl/site
    /usr/share/perl/site
    /usr/lib/perl/core
    /usr/share/perl/core
```

### PR DESCRIPTION
**Pull request by haarg**
Fixes #21315

### CODE REVIEW

**Comment by iabyn** [comment]
LGTM, except that Deparse's version needs bumping

**Comment by jkeenan** [comment]
I'm seeing this test failure in this branch:
```
$ cd t; ./perl harness ../lib/B/Deparse.t; cd -
../lib/B/Deparse.t .. 1/? # Failed test 289 - preserve __LINE__ etc at ../lib/B/Deparse.t line 96
#      got '{\n    CORE::state sub FOO () { 42 }\n    \n    use feature \'state\';\n    print 42, \"\\n\";\n}'
# expected /(?^:^\{\s*use\s+feature\s+\'state\'\;\s+state\s+sub\s+FOO\s+\:prototype\(\)\s+\{\s+42\s+\}\s+print\(FOO\,\s+\"\\n\"\)\;\s+\s*\}$)/
# =============================================
# CODE:
# --------
# use feature 'state';
# state sub FOO :prototype() { 42 }
# print(FOO, "\n");
# 
# --------
# EXPECTED:
# --------
# {
# use feature 'state';
# state sub FOO :prototype() { 42 }
# print(FOO, "\n");
# 
# }
# --------
# GOT:
# --------
# {\n    CORE::state sub FOO () { 42 }\n    \n    use feature \'state\';\n    print 42, \"\\n\";\n}
# --------
# =============================================
Attempt to free unreferenced scalar: SV 0x5b9097e7ff58 at ../lib/B/Deparse.t line 83, <DATA> chunk 288.
../lib/B/Deparse.t .. Failed 1/340 subtests 
	(less 2 skipped subtests: 337 okay)

Test Summary Report
-------------------
../lib/B/Deparse.t (Wstat: 0 Tests: 340 Failed: 1)
  Failed test:  289
Files=1, Tests=340,  1 wallclock secs ( 0.05 usr  0.00 sys +  1.23 cusr  0.15 csys =  1.43 CPU)
Result: FAIL
```
I think our CI is seeing that as well. @haarg, can you take a look?  thanks.

File: lib/B/Deparse.t [hunk 1]
```diff
@@ -3396,3 +3396,7 @@ $x //= exit 42;
 my $x = __LINE__;
 my $y = __FILE__;
 my $z = __PACKAGE__;
+####
+CORE::state sub FOO () { 42 }
+use feature 'state';
```
Comment (mauke) [inline lib/B/Deparse.t]: Why include this line if you're only using `CORE::state`?

**Comment by tonycoz** [comment]
> I think our CI is seeing that as well.

The CI failure is a cygwin issue, it looks like cygwin fork emulation isn't able to map a DLL where it was in the parent process.

File: lib/B/Deparse.t (hunk 1 above)
Comment (haarg) [reply lib/B/Deparse.t]: I was just matching the text in the test failure. I've fixed it so the `CORE::` isn't needed.

**Review by tonycoz — APPROVED** [review]

**Comment by iabyn** [comment]
The 'Attempt to free unreferenced scalar' message is now fixed by v5.39.9-66-g17535c984a.

**Pull request merged**